pkg/sql/virtual_table.go

```diff
@@ -24,20 +25,95 @@ import (
 // nil). If there is an error, then return (nil, error).
 type virtualTableGenerator func() (tree.Datums, error)
 
+type virtualTableGeneratorResponse struct {
+	datums tree.Datums
+	err    error
+}
+
+var errRowsDone = errors.New("predefined error that the generator has been canceled")
+
+// setupGenerator takes in a worker that generates rows eagerly and transforms
+// it into a lazy row generator. It returns three functions:
+// * cleanup: Performs all cleanup. May only be called once.
+// * next: A handle that can be called to generate a row from the worker. Next
+//   cannot be called once cleanup has been called.
+func setupGenerator(
+	worker func(addRow func(...tree.Datum) error) error,
+) (cleanup func(), next virtualTableGenerator) {
+	allowWork, done, send := make(chan struct{}), make(chan struct{}), make(chan virtualTableGeneratorResponse)
+	cleanup = func() {
+		close(done)
+	}
+	addRow := func(datums ...tree.Datum) error {
+		select {
+		case <-done:
```

Comment:
i think we should pass ctx in here for <-ctx.Done() as well

Reply:
I'm not sure if this is ok -- I don't know if the context passed into generate is the same one that can be canceled when we start making calls to next.

Reply:
hmm, maybe next or addRow should take in a context instead?